### ISSUE TITLE
ligand.pt is reused across the halogen_detail change, so every metal in a stale cache is read as chlorine

### ISSUE THREAD

**Issue by kuanlinhuang** [Huang-lab/TRACE-Kin#10]
`c5d76c7` ("ablate if ligand can benifit from more fine grained embedding of Metals and Helagons") flipped `smiles2graph`'s `halogen_detail` from `False` to `True`. That changes the size *and the meaning* of the atom-type vocabulary.

`ligand.pt` is keyed on nothing but its own existence:

```python
# training/train_trace_kin.py, preprocess()
if os.path.exists(ligand_path) and not force_rebuild:
    print(f"Reusing ligand cache: {ligand_path}")
    ligand_dict = torch.load(ligand_path)
```

and the LSF scripts say, in a comment, that reusing it is the intended path:

```
# No --force_rebuild: reuse caches across seeds.
```

So a run started after `c5d76c7` against a datafolder that already has a `ligand.pt` trains on the old vocabulary, silently. The ablation compares the cache to itself.

## What the stale codes mean

`mol_simplified_feature` stores `ATOM_CODES[atomic_num] + 1`, and `LigandEncoder.forward` feeds that straight into `self.atom_type_encoder = nn.Embedding(20, d_model)`. `nn.Embedding(20, …)` has room for both vocabularies, so nothing raises - the indices are simply reinterpreted:

| `x` in a cached graph | built with `halogen_detail=False` | read as, under `True` |
|---|---|---|
| 1-7 | B, C, N, O, P, S, Se | same |
| **8** | **any halogen** | **F** |
| **9** | **metal** | **Cl** |
| 10, 11 | *(unused)* | Br, I |
| 12 | *(unused)* | metal |

Per element:

```
  Cl (Z= 17): cached x= 8 (halogen) -> now means F     | a fresh build would write x=9
  Br (Z= 35): cached x= 8 (halogen) -> now means F     | a fresh build would write x=10
   I (Z= 53): cached x= 8 (halogen) -> now means F     | a fresh build would write x=11
  Zn (Z= 30): cached x= 9 (metal)   -> now means Cl    | a fresh build would write x=12
  Fe (Z= 26): cached x= 9 (metal)   -> now means Cl    | a fresh build would write x=12
```

**Every metal in a stale cache is read as chlorine**, and every Cl/Br/I as fluorine. Those are not neutral relabellings for a kinase-ligand model: the metal class is where zinc-binding-group chemistry lives, and the Cl→F confusion is precisely the halogen distinction the ablation was designed to test.

Two ways it goes wrong, and neither raises:

- **Cache present.** The "with halogen detail" arm trains on the "without" vocabulary. The ablation measures nothing, and reports a difference of zero as evidence that halogen detail does not help.
- **Caches built at different times across seeds.** The LSF comment describes seed 1 building and seeds 2 and 3 reusing. If seed 1's build straddles the flip, seeds disagree about what an atom is, and the seed variance absorbs it.

## Why the existing guard does not catch it

`preprocess()` already handles one version skew, and handles it well:

```python
missing_chembert = [s for s in ligand_dict if 'chembert_fp' not in ligand_dict[s]]
if missing_chembert:
    print(f"Backfilling chembert_fp on {len(missing_chembert)} cached ligands ...")
```

That works because the ChemBERT cutover *added a key*, so its absence is detectable. `halogen_detail` changes values under keys that were already there. The cache is structurally identical before and after; only the integers differ, and every one of them is in range.

## Suggested fix

Write the featurization parameters into the cache and refuse a mismatch, the way `fastVEP`'s transcript cache carries a format magic:

```python
FEATURIZATION = {"halogen_detail": True, "atom_feature": "atom_full_feature", "version": 2}

# on build
torch.save({"featurization": FEATURIZATION, "ligands": ligand_dict}, ligand_path)

# on load
blob = torch.load(ligand_path)
if blob.get("featurization") != FEATURIZATION:
    raise RuntimeError(
        f"{ligand_path} was built with {blob.get('featurization')}, this run needs "
        f"{FEATURIZATION}. Rerun with --force_rebuild."
    )
```

A dict comparison is enough; the point is that the file states what it is, so "reuse the cache" stops meaning "assume the code has not changed since". `protein.pt` and `molformer_emb.pt` have the same shape of problem and would want the same treatment.

Cheaper stopgap, if a format change is too much right now: delete every `ligand.pt` under the datafolders, and add `--force_rebuild` to the LSF scripts for the halogen arm specifically. That gets the current ablation right without touching the format, and leaves this open for the next featurization change.

## Also worth deciding

`halogen_detail=True` is hard-coded in `smiles2graph`, so it is not a config key and not recorded per run. It has the same shape as `mol_pe_mode` in #5: an option that decides what the model sees, whose value lives only in a Python literal. If it is a real ablation arm it belongs in `config_v7.json`, which would also give the cache key something to read.

*Filed from the 2026-09-02 Huang Lab code review. Related: #5, and [#9](https://github.com/Huang-lab/TRACE-Kin/pull/9) which adds the config-reachability checks for the PE options.*

**Comment by AtlasMaroc** [comment]
Hello Kuan,

I was indeed aware of how changing the codebase and ligand encoder will change the cache, thus I always deleted previous caches and forced rebuilt a new one.
